### ISSUE TITLE
# Threading Issue: Messages sent from non-platform thread cause crashes on Windows

### ISSUE THREAD

**Issue by SmokerDoker** [csdcorp/speech_to_text#645]
## Environment
- **OS**: Windows 11 (Standard PC, no VM)
- **Flutter**: 3.38.3 (Stable Channel)
- **Dart**: 3.10.1
- **speech_to_text**: ^7.0.0
- **speech_to_text_windows**: 1.0.0+beta.8 (transitive dependency)
- **Device**: Regular PC
- **Detection Language**: German

## Problem Description

When using the `speech_to_text` plugin on Windows, the native plugin sends platform channel messages from a background recognition thread instead of the platform (UI) thread. This violates Flutter's threading requirements and causes:

1. **Threading Warning** (always):
```
[ERROR:flutter/shell/common/shell.cc(1178)] The 'speech_to_text_windows' channel sent a message from native to Flutter on a non-platform thread. Platform channel messages must be sent on the platform thread. Failure to do so may result in data loss or crashes, and must be fixed in the plugin or application code creating that channel.
See https://docs.flutter.dev/platform-integration/platform-channels#channels-and-platform-threading for more information.
```

2. **Type Cast Error** (intermittent):
```
type 'Null' is not a subtype of type 'List<dynamic>' in type cast
```

3. **Silent Data Loss**: Speech recognition results are sometimes dropped, and the `onResult` callback is never invoked in Flutter/Dart code.

### 1. Setup `pubspec.yaml`
```yaml
dependencies:
  speech_to_text: ^7.0.0
```

### 2. Implement Speech Recognition Service (Correct Usage)

```dart
import 'package:flutter/foundation.dart';
import 'package:speech_to_text/speech_to_text.dart' as stt;

class SpeechRecognitionService extends ChangeNotifier {
  final stt.SpeechToText _speech = stt.SpeechToText();
  bool _isAvailable = false;
  bool _isListening = false;
  String _lastWords = '';

  bool get isAvailable => _isAvailable;
  bool get isListening => _isListening;
  String get lastWords => _lastWords;

  Future<void> initialize() async {
    try {
      _isAvailable = await _speech.initialize(
        onStatus: (status) {
          debugPrint('🎤 Speech Status: $status');
          if (status == 'done' || status == 'notListening') {
            _isListening = false;
            notifyListeners();
          }
        },
        onError: (error) {
          debugPrint('❌ Speech Error: $error');
          _isListening = false;
          notifyListeners();
        },
      );

      if (_isAvailable) {
        debugPrint('✅ Speech Recognition available');
      } else {
        debugPrint('❌ Speech Recognition not available');
      }
      
      notifyListeners();
    } catch (e) {
      debugPrint('❌ Error initializing speech: $e');
      _isAvailable = false;
      notifyListeners();
    }
  }

  Future<void> startListening() async {
    if (!_isAvailable || _isListening) return;

    _lastWords = '';
    _isListening = true;
    notifyListeners();

    await _speech.listen(
      onResult: _onSpeechResult,
      localeId: 'de_DE', // German locale
    );
  }

  Future<void> stopListening() async {
    if (!_isListening) return;
    
    await _speech.stop();
    _isListening = false;
    notifyListeners();
  }

  void _onSpeechResult(stt.SpeechRecognitionResult result) {
    // ⚠️ THIS CALLBACK IS NEVER CALLED due to threading issue!
    debugPrint('🎤 Recognized: "${result.recognizedWords}"');
    _lastWords = result.recognizedWords;
    notifyListeners();
  }

  @override
  void dispose() {
    _speech.stop();
    super.dispose();
  }
}
```

### 3. Run the app

```bash
flutter clean
flutter run -d windows
```

### 4. Trigger Speech Recognition

Start listening and speak into the microphone (e.g., say "Kobold" in German).

### Console Output:
```
Method called: listen
Starting speech recognition...
Sending status: listening
🎤 Speech Status: listening
Speech recognition listening started!
Recognition thread started
Speech event received. Event ID: 38
SPEI_RECOGNITION event
Recognized text: Kobold
Sending to Flutter (PATCHED): {"recognizedWords":"Kobold","finalResult":true,"alternates":[]}
[ERROR:flutter/shell/common/shell.cc(1178)] The 'speech_to_text_windows' channel sent a message from native to Flutter on a non-platform thread. Platform channel messages must be sent on the platform thread. Failure to do so may result in data loss or crashes, and must be fixed in the plugin or application code creating that channel.
See https://docs.flutter.dev/platform-integration/platform-channels#channels-and-platform-threading for more information.
Method called: stop
Stopping speech recognition...
Sending status: notListening
🎤 Speech Status: notListening
Speech recognition stopped
Recognition thread ended
```

### Key Observations:
1. ✅ Native code recognizes speech: `Recognized text: Kobold`
2. ✅ Native code sends to Flutter: `Sending to Flutter (PATCHED): {...}`
3. ❌ **Threading error appears**
4. ❌ **Dart callback `_onSpeechResult` is NEVER called** (no "🎤 Recognized:" log)
5. ❌ Flutter silently drops the message due to threading violation

## Root Cause Analysis

Looking at `speech_to_text_windows_plugin.cpp`, the issue is in the `Listen()` method:

```cpp
void SpeechToTextWindowsPlugin::Listen(...) {
  // Line 223: Creates a DETACHED background thread
  std::thread([this, ...] {
    // ... SAPI recognition loop ...
    
    while (m_listening) {
      SPEVENT event;
      HRESULT hr = m_cpRecoContext->GetEvents(1, &event, &eventsFetched);
      
      if (event.eEventId == SPEI_RECOGNITION) {
        // ... extract recognized text ...
        
        // ❌ PROBLEM: Called from background thread!
        SendTextRecognition(text, is_final);
      }
    }
  }).detach();  // ⚠️ Detached thread = not on UI thread
}

void SpeechToTextWindowsPlugin::SendTextRecognition(const std::string& text, bool is_final) {
  if (m_channel) {
    std::string json_result = "{\"recognizedWords\":\"" + text + 
                             "\",\"finalResult\":" + (is_final ? "true" : "false") + 
                             ",\"alternates\":[]}";
    
    // ❌ PROBLEM: InvokeMethod called from Recognition thread, not Platform thread!
    m_channel->InvokeMethod("textRecognition", 
        std::make_unique<flutter::EncodableValue>(json_result));
  }
}
```

**The problem**: `InvokeMethod` is called from the detached recognition thread, not from Flutter's platform (UI) thread.

## Expected Behavior

Messages should be dispatched to the platform thread before calling `InvokeMethod`, similar to how other platforms handle this:

```cpp
// Pseudo-code for proper fix
void SpeechToTextWindowsPlugin::SendTextRecognition(const std::string& text, bool is_final) {
  // Dispatch to UI thread first
  PostUIThreadTask([this, text, is_final]() {
    if (m_channel) {
      std::string json_result = "{\"recognizedWords\":\"" + text + 
                               "\",\"finalResult\":" + (is_final ? "true" : "false") + 
                               ",\"alternates\":[]}";
      m_channel->InvokeMethod("textRecognition", 
          std::make_unique<flutter::EncodableValue>(json_result));
    }
  });
}
```

## Workaround Applied

I temporarily patched the plugin to add the missing `"alternates":[]` field, which prevents the type cast error, but the threading issue remains and causes silent data loss.

## Impact

- ⚠️ **High**: Speech recognition callbacks don't reach Flutter/Dart code
- ⚠️ **High**: Silent data loss makes voice features unreliable
- ⚠️ **Medium**: Violates Flutter platform channel threading requirements
- ⚠️ **Low**: Type cast error (can be fixed separately by adding alternates field)

## Proposed Solution

Use Windows message queue or Flutter's task runner to dispatch messages to the platform thread:

1. **Option A**: Use `PostMessage` to UI thread with custom Windows message
2. **Option B**: Use Flutter's `TaskRunner` API to post to platform thread
3. **Option C**: Use `std::packaged_task` with platform thread executor

I'm happy to contribute a PR with the fix if you can provide guidance on the preferred approach for this plugin!

## Additional Notes

- This issue affects all Windows users of `speech_to_text`
- The plugin is currently in beta (1.0.0+beta.8), so this is a good time to fix it
- Other platform implementations (Android, iOS) handle threading correctly
- Flutter's documentation clearly states platform channel messages must be on platform thread

---

**Related Flutter Documentation**: https://docs.flutter.dev/platform-integration/platform-channels#channels-and-platform-threading
